lib/ignore/ignore_test.go

```diff
@@ -67,35 +67,85 @@ func TestExcludes(t *testing.T) {
 	!ign2
 	`
 	pats := New(true)
-	err := pats.Parse(bytes.NewBufferString(stignore), ".stignore")
+	err := pats.Parse(bytes.NewBufferString(stignore), ".stignore", "")
 	if err != nil {
 		t.Fatal(err)
 	}
 
 	var tests = []struct {
 		f string
-		r bool
+		r Result
 	}{
-		{"ign1", true},
-		{"ign2", true},
-		{"ibla2", true},
-		{"iex2", false},
-		{filepath.Join("ign1", "ign"), true},
-		{filepath.Join("ign1", "ex"), false},
-		{filepath.Join("ign1", "iex2"), false},
-		{filepath.Join("iex2", "ign"), false},
-		{filepath.Join("foo", "bar", "ign1"), true},
-		{filepath.Join("foo", "bar", "ign2"), true},
-		{filepath.Join("foo", "bar", "iex2"), false},
+		{"ign1", Nuke},
+		{"ign2", Nuke},
+		{"ibla2", Nuke},
+		{"iex2", DontIgnore},
+		{filepath.Join("ign1", "ign"), Nuke},
+		{filepath.Join("ign1", "ex"), DontIgnore},
+		{filepath.Join("ign1", "iex2"), DontIgnore},
+		{filepath.Join("iex2", "ign"), DontIgnore},
+		{filepath.Join("foo", "bar", "ign1"), Nuke},
+		{filepath.Join("foo", "bar", "ign2"), Nuke},
+		{filepath.Join("foo", "bar", "iex2"), DontIgnore},
 	}
 
 	for _, tc := range tests {
-		if r := pats.Match(tc.f); r != tc.r {
+		if r := pats.match(tc.f); r != tc.r {
 			t.Errorf("Incorrect match for %s: %v != %v", tc.f, r, tc.r)
 		}
 	}
 }
 
+func TestPreserve(t *testing.T) {
+	stignore := `
+	(?preserve)iex2
+	(?preserve)Path with Spaces
+	!(?preserve)ign1/ex
+	ign1
+	!i*2
+	!(?preserve)ign2
+	(?preserve)(?i)**ifn1
+	!(?preserve)(?i)**ifn2
+	!(?preserve)(?i)**ifn3
+	(?preserve)(?i)ifn4
+	`
+	pats := New(true)
+	err := pats.Parse(bytes.NewBufferString(stignore), ".stignore", "")
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	var tests = []struct {
+		f string
+		r Result
+	}{
+		{"ign1", Nuke},
+		{"Path with Spaces", Preserve},
+		{"ign2", DontIgnore},
+		{"ibla2", DontIgnore},
+		{"iex2", Preserve},
+		{filepath.Join("ign1", "ign"), Nuke},
+		{filepath.Join("ign1", "ex"), Preserve},
```

Comment:
How does this become `Preserve` - shouldn't it be `DontIgnore`? Or, more generally, what is `!(?preserve)...` supposed to mean exactly, in English? :D